### ISSUE TITLE
When does patch version get reset to 0 or 1?

### ISSUE THREAD

**Issue by johnwc** [dotnet/Nerdbank.GitVersioning#1117]
When using this tool, and creating a release branch using `prepare-release` the patch number does not get reset back to 0 or 1 in the release branch. How do we get it to reset, so that releases start at `v4.3.0` instead of `v4.3.1409`?

**Comment by AArnott** [comment]
What you describe isn't the default behavior, because that is discouraged behavior. Doing that would mean you're building the same package version for multiple distinct versions of your source code. That leads to package cache poisoning, where someone who already had v4.3.1 from your main branch in their package cache cannot consume v4.3.1 from your release branch until they realize they have to delete the old 4.3.1 package from their cache so that nuget restore brings down the new one. 
And if your CI pushes packages from main to a nuget feed, you simply won't have the option to push 4.3.1 a second time.

So while you *can* do what you want (by setting `versionHeightOffset: -1409` in your version.json file), I strongly discourage it.

And FWIW, I've never heard a customer complain about 4.3.1409 being released instead of 4.3.0 or 4.3.1. A very few developers of the software itself (such as yourself) feel a bit ruffled at first when they discover this behavior, but I encourage you to consider giving up the need to release a .0 version. I haven't in many years... and no user has even mentioned that they noticed.

**Comment by johnwc** [comment]
We never have the issues that you refer to because our main branch is always labeled as being a prerelease, which includes beta/alpha and the commit id. So, never causes two builds to have the same version. Release branches under `releases/` are the only branches that get a stable release stamped on them as well as only version that can make to production.

Also, a lot of people make the claim `I never heard of...` so it must not be important. I would say that a lot of people don't stand up and voice their opinion or requirements and just fold to what cards they've been handed. Just because we don't hear about things, does not mean they are not important to someone else. A very good scenario for wanting the revision reset to 0 or 1, is the KISS principle. Keep it simple stupid. Our team along with our users want to be able to look at a release version and know quickly if a new version was pushed. Being able to see `v4.3.8` move to `v4.3.9` or `v4.3.10` is much easier to remember, than `v4.3.1423` moving to `v4.3.1432` (Think about someone looking at the version on Friday as they leave their office, knowing there is a fix being worked on they need. Then coming in Monday to see if the version updated over the weekend.)

**Comment by johnwc** [comment]
We do that, we set the height after the prerelease like in your example. This was the only way we could get npm to see the next prerelease as an update to a npm package. We originally had it doing the OoB version stamping of `v1.0.40-beta`, releasing it to our internal npm repo and using it within other packages. But when the next update came out and it was now `v1.0.41-beta`, `npm outdated` did not see that the new update was in fact 'wanted'. Once we changed it to `v1.0-beta.40`, we could simple run `npm update` and it would update to next prerelease.

As for the consecutive patch versions in release branches. When we need to make a patch to a release, we create a new branch from the release, work on patch, make a PR from patch branch back into release branch using squash merge. Causing all the commits from patch to be in one commit. Then we create a PR from patch branch into main, using standard marge that keeps commit history. Our main branch has all history from beginning of project, allowing to look for specific work or changes if needed.

We would much rather not have to manually set the third digit, or manually apply the `versionHeightOffset` in the version.json. Can there be a new command line switch for prepare-release that has it set the `versionHeightOffset` to current height in the release branch?

**Comment by AArnott** [comment]
> Can there be a new command line switch for prepare-release that has it set the versionHeightOffset to current height in the release branch?

I don't see why not. I think the switch should fail if the version height had not already been moved into the prerelease tag as discussed above to avoid inadvertently reusing numbers or violating semver ordering.

Would you care to author that pull request?

**Comment by johnwc** [comment]
Can you explain the constraint better, how do we test for that situation?

I wouldn't mind authoring the addition.

**Comment by johnwc** [comment]
Ok, so do the following verifiactions

1. A regex test on the version to make sure the height is after the prerelease tag. 
2. If they set a tag in the command line, be sure it sorts after whatever the prerelease tag is set via the version.json.

If both pass, set the `versionHeightOffset` field on the release branches version.json. If `versionHeightOffset` already has a value, update it.

**Comment by AArnott** [comment]
> If both pass

*and* the CLI switch requesting this behavior is present.

> A regex test on the version

I suppose, if that's the easiest way to go about it. But I suspect when you get into that code that you'll find a simpler way. For example just the presence of `{height}` might infer that there's a prerelease tag.

**Comment by johnwc** [comment]
> I suppose, if that's the easiest way to go about it. But I suspect when you get into that code that you'll find a simpler way. For example just the presence of {height} might infer that there's a prerelease tag.

I can just do a simple check that version ends with `{height}` instead of the regex test.

**Comment by johnwc** [comment]
@AArnott unrelated to this, but what would cause this? Running `nbgv cloud -a` outputs that the version is `1.3.1-beta.5...` but the build gets set as `1.3.1-beta.0...` And the library project that was updated was stamped with `1.3.1-beta.3...` We are using the [mono repo versioning](https://github.com/dotnet/Nerdbank.GitVersioning/blob/main/doc/pathFilters.md) to only version the project that was changed, just to give some perspective of the setup.

![Image](https://github.com/user-attachments/assets/5f2c97cd-07d9-437c-965f-93e7a71f8018)

![Image](https://github.com/user-attachments/assets/3462145b-3c2e-423d-bea0-e10fc2f6c540)

**Comment by AArnott** [comment]
`nbgv` doesn't consider msbuild, so the only thing that matters is what directory it considers as the base, which is the working directory by default.
But msbuild defaults to considering the project directory as the input to version calculation. So if that isn't the same as where you run `nbgv`, that can explain the difference.
Also, an msbuild property exists that tells a project to always use some _other_ directory as the basis for version calculation. So if you're setting that, you may want to try running `nbgv` from that directory for consistent results.

**Comment by johnwc** [comment]
But what about the inconsistency of the build label not being what is printed in the task output from `nbgv cloud -a`?

**Comment by AArnott** [comment]
Oh that. It's probably that you're setting the build label to something else later in the build. In fact the default behavior of Nerdbank.GitVersioning is to set the cloud build number for every msbuild project's build, which in your case probably doesn't make sense and you should probably [turn that off](https://github.com/AArnott/Library.Template/blob/373a8ad2d2da268a7879dc937a2377a7b0b50786/version.json#L8-L10).

**Comment by nabeelio** [comment]
> When using this tool, and creating a release branch using `prepare-release` the patch number does not get reset back to 0 or 1 in the release branch. How do we get it to reset, so that releases start at `v4.3.0` instead of `v4.3.1409`?

It seems like it's reset for me and I'm not quite sure why. I was at a height of ~600, and I was messing with the version, (at 2.0, set it to 2.0-dev, to see if it would go to 2.0.601-dev), it didn't do that, and then removed it, but didn't revert any of those commits. I read in another ticket that I can't find now, if that had been done, it resets the height... is that true?

**Comment by AArnott** [comment]
No, moving from `2.0` to `2.0-dev` does not reset the version height. Nor does the reverse direction. 

@nabeelio I'm guessing you're thinking of https://github.com/dotnet/Nerdbank.GitVersioning/issues/1185#issuecomment-2687895595.

**Comment by johnwc** [comment]
@AArnott we've actually seen this exact scenario occur to us as well. simple editing the version.json in a branch causes the height to reset. Resting the change so to remove the commit and revert the file causes it to go back to the original height. We just thought this was how it was designed, did not know it is not supposed to work like that.

**Comment by AArnott** [comment]
@johnwc please send exact repro steps, or share a git repo itself with this in its history and I'll check it out.

**Comment by johnwc** [comment]
@AArnott unfortunately it occurs so rarely, it would be difficult to for us to create a reproducible public repo. If you give me things to look for or command to run and share back, when it occurs again in one of our private repos, I can do that.

**Comment by johnwc** [comment]
Maybe have an opt-in option that can be set in the config, that is default set for new installs? Everyone else can choose to make the switch on their own after reading the side effects.

**Comment by AArnott** [comment]
@johnwc I've never heard of and can't imagine a git situation where adding a commit would not increment the calculated version height. Are you sure adding commits doesn't increase it?
Or rather is the problem that the height appeared to go backwards, and though it increases with each commit, you need it to be much higher than it currently is?

**Comment by AArnott** [comment]
> until I completed the PR

What was the state before you completed the PR? Were you testing version calculation on a local machine or was the version evident in a PR build?
In general, if your PR brings together a branch with version 1.1.10 and another branch that would build at 1.1.30, I would expect the PR build to build as 1.1.31. 

This is because NB.GV always operates on the HEAD commit, and for a PR build that would be the candidate merge commit. And NB.GV, for a commit with multiple parents, will always take the higher version height of the multiple parent paths, and then add 1 to that version height.
Now, if one of those parents also bumped the major or minor version in version.json and the other didn't, NB.GV will take the merge result's version number (which will typically be the higher of the two unless you did something funky with the merge commit) and then assess the version height from when that version was established. That would practically speaking mean it would just add +1 to the version height from the parent that already had the matching major.minor version.

**Comment by AArnott** [comment]
Closing due to no response.
